# 12 In-Class Assignment: Change of Basis

<img alt="Alternative visual representation showing how the four Fundamental Subspaces spaces map to each other" src="https://upload.wikimedia.org/wikipedia/commons/4/4c/KerIm_2015Joz_L2.png" width="50%">

Image from: [wikipedia](https://en.wikipedia.org/wiki/Kernel_(linear_algebra))


### Agenda for today's class (80 minutes)

1. [(20 minutes) Review Pre-class assignment](#Review_Pre-class_assignment)
1. [(20 minutes) The Four Fundamental Spaces](#The_Four_Fundamental_Spaces)
1. [(20 minutes) Matrix Representation of Vector Spaces](#Matrix_Representation_of_Vector_Spaces)
1. [(20 minutes) Practice Example](#Practice_Example)

In [1]:
%matplotlib inline
import matplotlib.pylab as plt
import numpy as np
import sympy as sym
sym.init_printing(use_unicode=True)

----
<a name="Review_Pre-class_assignment"></a>
## 1. Review Pre-class assignment


--
<a name="The_Four_Fundamental_Spaces"></a>

## 2. The Four Fundamental Spaces - Instructor Led

In this course we represent a system of linear equations as $Ax=b$. 
The matrix $A$ can be viewed as taking a point $x$ in the input space and projecting that point to $b$ in the output space.  


It turns out, everything we need to know about $A$ is represented by four fundamental vector spaces.  Two of the four spaces are easily defined as follows:

> *Row space of $A$*: linear combination of all rows of $A$

> *Column space of $A$*: linear combination of all columns of $A$

The other two fundamental spaces are defined by a concept called the *Null Space*. 
The *Null space* is calculated by finding all the solutions to the homogeneous system $Ax=0$. The final two fundamental spaces are defined as follows:

> *Null space or kernel of $A$*: all $x$ such that $Ax=0$

> *Null space of $A^\top$*: all $y$ such that $A^\top y =0$

<img alt="Classic picture of the four fundamental spaces. Please see text for detailed description." src="https://kevinbinz.files.wordpress.com/2017/02/linear-algebra-fundamental-space-interpretation-6.png" width="100%">

Image from: https://kevinbinz.com/2017/02/20/linear-algebra/

### The four fundamental subspaces

* Columnspace, $\mathcal{C}(A)$
* Nullspace, $\mathcal{N}(A)$
* Rowspaces, $R(A)$
    * All linear combinations of rows
    * All the linear combinations of the colums of $A^\top$, $\mathcal{C}(A^\top)$
* Nullspace of $A^\top$, $\mathcal{N}(A^\top)$ (the left nullspace of $A$)

----
<a name="Matrix_Representation_of_Vector_Spaces"></a>
## 3. Matrix Representation of Vector Spaces

Consider the following matrix $A$. 

$$ 
\left[
\begin{matrix}
    1 & 0 & 3  \\
    0 & 1 & 5  \\
    1 & 1 & 8 
\end{matrix}
\right] 
$$

&#9989; **<font color=red>QUESTION:</font>** What is the reduced row echelon form of $A$?

In [2]:
# Put your answer to the above question here.
A = np.matrix([[1,0,3],[0,1,5],[1,1,8]])
rref = sym.Matrix(A).rref()[0]
rref

⎡1  0  3⎤
⎢       ⎥
⎢0  1  5⎥
⎢       ⎥
⎣0  0  0⎦

In [3]:
from answercheck import checkanswer

checkanswer.matrix(rref,'1731818a1555cc33a778a4eb76af945c');


    Trying to convert to a array matrix using ```A = np.matrix(A)```.


    Trying to convert to float using ```A = A.astype(float)```.


    Converting to positive values of zero using  ```A[A==-0] = 0```.

Testing [[1. 0. 3.]
 [0. 1. 5.]
 [0. 0. 0.]]
Answer seems to be correct



**ROW SPACE** The first and second (non zero) rows of the above matrix "spans" the same space as the orignal three row vectors in $A$. We often call this the "row space" and it can be written as a linear combination of the non-zero rows of the reduced row echelon form:

$$row(A) = \{r(1,0,3)^\top+s(0,1,5)^\top : r,s \in \mathbb{R} \} = \mbox{span}((1,0,3)^\top,(0,1,5)^\top)$$

&#9989; **<font color=red>DO THIS:</font>** Calculate  the solutions to the system of homogeneous equations $Ax=0$. This is often called the **NULL SPACE** or sometimes **KERNEL** of $A$.

In [4]:
zero = np.matrix([[0,0,0]]).T
answer = np.linalg.solve(A,zero)
print(answer)

LinAlgError: Singular matrix

${z} \times {\left[
\begin{matrix}
    -3 \\
    -5  \\
    1 
\end{matrix}
\right] }$

In [5]:
sym.Matrix(A).nullspace()

⎡⎡-3⎤⎤
⎢⎢  ⎥⎥
⎢⎢-5⎥⎥
⎢⎢  ⎥⎥
⎣⎣1 ⎦⎦

&#9989; **<font color=red>DO THIS:</font>** We introduced two subspaces of $\mathbb R^3$. 
Pick one vector from the row space of $A$ and another vector from the null space of $A$. 
Find the dot product of these two vectors.

In [6]:
#Put your answer here
s = np.array([0,1,5])
r = np.array([1,0,3])
r_s = 3*r + 2*s
null = np.array([-3,-5,1])
answer = np.dot(r_s, null)
print(answer)

0


&#9989; **<font color=red>Question:</font>** Did you get the same value for the dot product as your neighbor (or if you are working by yourself, did you get the same value for the dot product if you made different choices for the representatives of the row space and null space of $A$)? Explain your answer. 

**Yes, it was because the dot product of row space and null space is 0**

&#9989; **<font color=red>DO THIS:</font>** What is the reduced row echelon form of $A^T$?

In [7]:
#Put your answer here
At = np.matrix(A).T
rref = sym.Matrix(At).rref()[0]
rref

⎡1  0  1⎤
⎢       ⎥
⎢0  1  1⎥
⎢       ⎥
⎣0  0  0⎦

**COLUMN SPACE:** The first and second (non zero) rows of the above matrix "spans" the same space as the original three column vectors in $A$. We often call this the "column space" (or "image space") of $A$ and it can be written as a linear combination of the non-zero rows of the reduced row echelon form of $A^T$:

$$col(A) = \{a(1,0,1)^\top+b(0,1,1)^\top : a,b \in \mathbb{R} \} = \mbox{span}((1,0,1)^\top,(0,1,1)^\top)$$

&#9989; **<font color=red>DO THIS:</font>** Calculate the solutions to the system of homogeneous equations $A^Tx=0$. This is often called the **NULL SPACE** of $A^T$.

In [8]:
sym.Matrix(At).nullspace()

⎡⎡-1⎤⎤
⎢⎢  ⎥⎥
⎢⎢-1⎥⎥
⎢⎢  ⎥⎥
⎣⎣1 ⎦⎦

**Note:** The fact that you got zero as the dot product for a vector from the row space and a vector from the null space was not a coincidence!

**<font color=red>Theorem:</font>** *Every vector $x$ in the null space of $A$ is perpendicular to every row of $A$, because $Ax=0$. The Null space and the Row space are orthogonal subspaces of $\mathbb R^n$.*

*Every vector $y$ in the null space of $A^T$ is perpendicular to every column of $A$, because $A^Ty=0$.*

---
### Example #1 

Consider the following system of linear equations.

$$ x_1 - x_2 + x_3 = 3 $$
$$ -2x_1 + 2x_2 - 2x_3 = -6 $$

&#9989; **<font color=red>DO THIS:</font>** What are the solutions to the above system of equations?

In [9]:
# Put your code here
A = np.matrix([[1,-1,1,3],[-2,2,-2,-6]])
rref = sym.Matrix(A).rref()[0]
rref

⎡1  -1  1  3⎤
⎢           ⎥
⎣0  0   0  0⎦

Therefore, the solution would be
$$ x-y+z = 3$$
$$y = y$$
$$z=z$$

&#9989; **<font color=red>DO THIS:</font>** Come up with a specific arbitrary solution (any solution will do) to the above set of equations.

**x = 3, y = 1, z = 1**

&#9989; **<font color=red>DO THIS:</font>** Now consider only the left hand side of the above matrix and solve for the kernel (null Space) of A:


$$ A = 
\left[
\begin{matrix}
    1 & -1 & 1  \\
    -2 & 2 & -2  
\end{matrix}
\right] 
$$

In [10]:
#Put your answer here
A = np.matrix([[1,-1,1],[-2,2,-2]])
sym.Matrix(A).nullspace()

⎡⎡1⎤  ⎡-1⎤⎤
⎢⎢ ⎥  ⎢  ⎥⎥
⎢⎢1⎥, ⎢0 ⎥⎥
⎢⎢ ⎥  ⎢  ⎥⎥
⎣⎣0⎦  ⎣1 ⎦⎦

&#9989; **<font color=red>DO THIS:</font>** Express an arbitrary solution as the sum of an element of the kernel of the transformation defined by the matrix of coefficients and a particular solution.

**Not able to do this. Would go to helproom and learn more about it.**

&#9989; **<font color=red>DO THIS:</font>** Discuss in your group and the class your solution from above. How does the solution to $Ax=b$ relate to the solution to $Ax=0$.  If you were to plot all solutions, what shape does it take? How does this shape relate to the kernel?

We were not able to discuss it as we were discussing other questions.

----
<a name="Practice_Example"></a>
## 3. Practice Nutrition

Big Annie's Mac and Cheese fans want to improve the levels of protein and fiber for lunch by adding broccoli and canned chicken. 
The nutrition information for the foods in this problem are 


|Nutrient    | Mac and Cheese           |  Broccoli        |    Chicken   | Shells and White Cheddar |
|----|-----------------|----------------|----------|----------|
|Calories| 270 | 51 |  70 | 260 |
|Protein (g) | 10 | 5.4 |  15| 9|
|Fiber (g)| 2   |  5.2 |  0| 5|


<img alt="Logo for Annie's Mac and Cheese" src="https://upload.wikimedia.org/wikipedia/commons/c/cd/Annies_logo.png" width="50%">

She wants to achieve the goals with exactly 400 calories, 30 g of protein, and 10 g of fiber by choosing the combination of these three or four servings. (Assume that we can have non-integer proportions for each serving.)

&#9989; **<font color=red>Question a</font>**: We consider all **four** choices of food together. Formulate the problem into a system of equations 
$$Ax = b.$$
Create your matrix $A$ and the column vector $b$ in *np.matrix*. 

In [11]:
import numpy as np
#####Start your code here #####
A = np.matrix([[270,51,70,260],[10,5.4,15,9],[2,5.2,0,5]])
b = np.matrix([[400,30,10]]).T
#####End of your code here#####

&#9989; **<font color=red>Question b</font>**: In this and next question, we only consider **three** out of the four choices. What proportions of these servings of the **three** foods (Mac and Cheese, Broccoli, and Chicken) should be used to meet the goal? (Hint: formulate it as a system of equations and solve it).

In [12]:
#Put your answer here
A = np.matrix([[270,51,70],[10,5.4,15],[2,5.2,0]])
b = np.matrix([[400,30,10]]).T
x1 = np.linalg.solve(A,b).T
sym.Matrix(x1)

[0.985822419800553  1.54391445392286  0.786975850054067]

&#9989; **<font color=red>Question c</font>**: She found that there was too much broccoli in the proportions from part (b), so she decided to switch from classical Mac and Cheese to Annie’s Whole Wheat Shells and White Cheddar. What proportions of servings of the new **three** food should she use to meet the goals?

In [13]:
#Put your answer here
A = np.matrix([[260,51,70],[9,5.4,15],[5,5.2,0]])
b = np.matrix([[400,30,10]]).T
x2 = np.linalg.solve(A,b).T
sym.Matrix(x2)

[1.08899064304201  0.875970535536532  1.03125622138164]

&#9989; **<font color=red>Question d</font>**: Based on the solutions to parts (b) and (c), what are the possible proportions of serving for the **four** food that meet the goal. 

In [14]:
x1 = np.matrix([[x1[0,0],x1[0,1],x1[0,2],0]]).T
x2 = np.matrix([[0,x2[0,0],x2[0,1],x2[0,2]]]).T
x = x1 -x2
sym.Matrix(x)

⎡ 0.985822419800553 ⎤
⎢                   ⎥
⎢ 0.454923810880857 ⎥
⎢                   ⎥
⎢-0.0889946854824646⎥
⎢                   ⎥
⎣ -1.03125622138164 ⎦

General solution :-


 $$\left[
\begin{matrix}
    0.98 \\ 1.54  \\0.78 \\0
 \end{matrix}
\right]
+
{c}\times{\left[
\begin{matrix}
    0.98 \\ 0.45\\-0.08\\-1.03  
 \end{matrix}
\right]}
$$



In [15]:
A = np.matrix([[5,4,3],[0,0,0],[0,0,0]]).T
B = np.matrix([[4,3,2],[0,0,0],[0,0,0]]).T
Bi = np.linalg.inv(B)
#answer = A*Bi
print(Bi)

LinAlgError: Singular matrix

&#9989; **<font color=red>Question e</font>**: Solve the system of equations from part (a). You need to first decide the three outcomes: One solution, None solution, Infinite many solutions. Then for *One solution*, write down the solution; for *Infinite many solutions*, write down all the solutions with free variables. 

In [16]:
#Put your answer here
A = np.matrix([[270,51,70,260,400],[10,5.4,15,9,30],[2,5.2,0,5,10]])
rref = sym.Matrix(A).rref()[0]
rref

⎡1  0  0  0.905262525531659   0.985822419800553⎤
⎢                                              ⎥
⎢0  1  0  0.613360567103208   1.54391445392286 ⎥
⎢                                              ⎥
⎣0  0  1  -0.224318154511594  0.786975850054067⎦

----

Written by Dr. Dirk Colbry, Michigan State University
<a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International License</a>.